In [1]:
import pandas as pd
import tensorflow as tf

## 데이터 준비

In [2]:
directory = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris = pd.read_csv(directory)

In [3]:
iris.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [6]:
iris = pd.get_dummies(iris)
iris.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2          1              0             0
1   4.9  3.0    1.4   0.2          1              0             0
2   4.7  3.2    1.3   0.2          1              0             0
3   4.6  3.1    1.5   0.2          1              0             0
4   5.0  3.6    1.4   0.2          1              0             0

In [7]:
ind_var = iris.iloc[:,:4]
dep_var = iris.iloc[:,4:]

In [9]:
print(ind_var.shape, dep_var.shape)

(150, 4) (150, 3)


## 모델의 구조 만들기

In [10]:
X = tf.keras.layers.Input(shape=[4])
Y = tf.keras.layers.Dense(3, activation='softmax')(X)
model = tf.keras.models.Model(X, Y)
model.compile(loss='categorical_crossentropy',
             metrics='accuracy')

## 모델 학습하기

In [11]:
model.fit(ind_var,
          dep_var,
          epochs=10)
model.fit(ind_var,
          dep_var,
          epochs=1000,
          verbose=0)
print()
model.fit(ind_var,
          dep_var,
          epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 998us/step - loss: 5.6837 - accuracy: 0.3333
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 5.5593 - accuracy: 0.3333
Epoch 3/10
5/5 [==============================] - 0s 998us/step - loss: 5.4766 - accuracy: 0.3333
Epoch 4/10
5/5 [==============================] - 0s 1ms/step - loss: 5.4037 - accuracy: 0.3333
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 5.3368 - accuracy: 0.3333
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 5.2755 - accuracy: 0.3333
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 5.2141 - accuracy: 0.3333
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 5.1582 - accuracy: 0.3333
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 5.1009 - accuracy: 0.3333
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 5.0437 - accuracy: 0.3333

Epoch 1/10
5/5 [==================

In [13]:
pd.DataFrame(model.predict(ind_var[:5]))

0         1             2
0  0.998919  0.001081  9.164519e-08
1  0.995615  0.004385  7.721193e-07
2  0.998002  0.001997  3.346266e-07
3  0.994637  0.005362  1.546767e-06
4  0.999135  0.000865  7.733865e-08

In [14]:
dep_var[:5]

품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0

## 모델의 수식

In [17]:
model.get_weights()

[array([[ 1.2120858 ,  0.79970855, -0.62333703],
        [ 1.92979   , -0.71278566, -1.605099  ],
        [-3.3558195 ,  0.07102513,  1.630443  ],
        [-3.9565172 , -1.8591785 ,  1.597882  ]], dtype=float32),
 array([ 1.4604897,  0.7665555, -1.1026498], dtype=float32)]

- 모델은 총 3개의 수식을 반환 (열에 따라)
- 예를 들어 setosa의 경우 $1.21*꽃잎길이 + 1.92*꽃잎폭 -3.36*꽃받침길이 - 3.96*꽃받침폭 + 1.46$으로 계산하게 됨